pkg/apps/apis/apps/validation/validation_test.go

```diff
@@ -911,54 +912,224 @@ func mkintp(i int) *int {
+		},
+		{
+			// Setting deployment: selector is invalid as it changes for every RS by LatestVersion
+			dc: &appsapi.DeploymentConfig{
+				ObjectMeta: metav1.ObjectMeta{Name: dcName},
+				Spec: appsapi.DeploymentConfigSpec{
+					Selector: map[string]string{
+						appsapi.DeploymentLabel: dcName,
+					},
+					Strategy: test.OkStrategy(),
+					Template: func() *kapi.PodTemplateSpec {
+						t := test.OkPodTemplate()
+						t.Labels = map[string]string{
+							appsapi.DeploymentLabel: dcName,
+						}
+						return t
+					}(),
+				},
+			},
+			expectedErrors: field.ErrorList{
+				&field.Error{
+					Type:   field.ErrorTypeInvalid,
+					Field:  "spec.selector",
+					Detail: "can't set restricted selector `deployment`",
+				},
+				&field.Error{
+					Type:   field.ErrorTypeInvalid,
+					Field:  "spec.template.labels",
+					Detail: "can't set restricted label `deployment`",
+				},
+			},
+		},
+		{
+			// Setting deployment: label is invalid as it would be overwritten
+			dc: &appsapi.DeploymentConfig{
+				ObjectMeta: metav1.ObjectMeta{Name: dcName},
+				Spec: appsapi.DeploymentConfigSpec{
+					Selector: map[string]string{
+						"foo": "bar",
+					},
+					Strategy: test.OkStrategy(),
+					Template: func() *kapi.PodTemplateSpec {
+						t := test.OkPodTemplate()
+						t.Labels = map[string]string{
+							"foo": "bar",
+							appsapi.DeploymentLabel: dcName,
+						}
+						return t
+					}(),
+				},
+			},
+			expectedErrors: field.ErrorList{
+				&field.Error{
+					Type:   field.ErrorTypeInvalid,
+					Field:  "spec.template.labels",
+					Detail: "can't set restricted label `deployment`",
+				},
 			},
-			expectedErr: true,
-			errorType:   field.ErrorTypeInvalid,
-			field:       "spec.template.metadata.labels",
 		},
 	}
 
-	for name, test := range tests {
-		if test.expectedErr {
-			test.spec.Template.Labels["a"] = "c"
-		}
-		errs := ValidateDeploymentConfigSpec(test.spec)
-		if len(errs) == 0 && test.expectedErr {
-			t.Errorf("%s: expected failure", name)
-			continue
-		}
-		if !test.expectedErr {
-			continue
-		}
-		if len(errs) != 1 {
-			t.Errorf("%s: expected one error, got %d", name, len(errs))
-			continue
-		}
-		err := errs[0]
-		if err.Type != test.errorType {
-			t.Errorf("%s: expected error to have type %q, got %q", name, test.errorType, err.Type)
-		}
-		if err.Field != test.field {
-			t.Errorf("%s: expected error to have field %q, got %q", name, test.field, err.Field)
-		}
+	for _, tc := range tt {
+		t.Run("", func(t *testing.T) {
+			t.Logf("dc.Spec.Selector: %#v", tc.dc.Spec.Selector)
+			t.Logf("dc.Spec.Template.Labels: %#v", tc.dc.Spec.Template.Labels)
+
+			errs := ValidateDeploymentConfigSpec(tc.dc)
+			if len(tc.expectedErrors) == 0 && len(errs) != 0 {
+				t.Errorf("expected no error but got: '%v'", errs.ToAggregate())
+				return
+			}
+
+			if len(tc.expectedErrors) != 0 && len(errs) == 0 {
+				t.Errorf("expected errors '%v', got none", tc.expectedErrors.ToAggregate())
+				return
+
```

Comment:
TODO: remove this white space if there are any comments to be addressed (don't want tore-run test just because of this change)